src/rest-server/src/templates/yarnContainerScript.mustache

```diff
@@ -26,44 +26,47 @@ export PAI_DEFAULT_FS_URI={{ hdfsUri }}
 exec 13>$LAUNCHER_LOG_DIR/YarnContainerDebug.log
 BASH_XTRACEFD=13
 
+# YARN kill container will send SIGTERM(15) and SIGKILL(9)
+function sigterm_handler()
+{
+  printf "[DEBUG] SIGTERM received, job may be killed by YARN\n"
+  killed_by_yarn=true
+  exit 143
+}
+
 
 function exit_handler()
 {
   rc=$?
-  echo "Exited with $rc"
-  local handler="Yarn container exit handler"
-  debug_log "$handler"  "EXIT signal received in yarn container, performing clean up action..."
-
-  debug_log "$handler" "trying to kill docker container $docker_name"
-  docker logs $docker_name
-  docker inspect $docker_name
-  pid=$(docker inspect --format={{{ inspectPidFormat }}} $docker_name 2>/dev/null)
-  if [ $pid -gt 0 ]; then
-    kill -9 $pid &&\
-      debug_log "$handler" "docker caontainer $docker_name killed successfully." ||\
-      debug_log "$handler" "tries to kill the container $docker_name but failed. Maybe it has already exited."
-  else
-    debug_log "$handler" "docker container $docker_name has already exited"
-    is_oom=$(docker inspect --format={{{ inspectOOMKilledFormat }}} $docker_name 2>/dev/null)
-    debug_log "$handler" "docker container $docker_name is exited by OOM? $is_oom"
-    if [ "$is_oom" = "true" ]; then
-      rc=55
-    fi
+  printf "[DEBUG] Performing clean up action...\n"
+  if [[ $rc -eq 125 ]]; then
+    # We assume 125 is a docker error, and retry it without burning user's retry count
+    printf "[DEBUG] Exit code 125, it may be docker error, convert it to a transient error\n"
+    rc=205
   fi
 
-  docker container rm $docker_name
-  pkill --parent $$
+  if [[ ${killed_by_yarn} != true && $rc -eq 143 || $rc -eq 137 ]]; then
+    printf "[DEBUG] Exit code $rc, it may be due to out of memory (OOM) or other critical errors.\n"
+    rc=55
+  fi
 
-  debug_log "$handler" "write exit code to file"
-  debug_log "$handler" "yarn container exit code: $rc"
-  debug_log "$handler" "exit code file path: /var/lib/hadoopdata/nm-local-dir/nmPrivate/$APP_ID/$CONTAINER_ID/$CONTAINER_ID.pid.exitcode"
+  printf "[DEBUG] Write exit code $rc to file /var/lib/hadoopdata/nm-local-dir/nmPrivate/$APP_ID/$CONTAINER_ID/$CONTAINER_ID.pid.exitcode.\n"
   echo $rc > "/var/lib/hadoopdata/nm-local-dir/nmPrivate/$APP_ID/$CONTAINER_ID/$CONTAINER_ID.pid.exitcode"
 
+  # After all necessary steps, do a best effort kill here
```

Comment:
what is the behavior when we stuck in (docker inspect), and yarn sends a SIGKILL to yarncontainer?
in this case, "/var/lib/hadoopdata/nm-local-dir/nmPrivate/$APP_ID/$CONTAINER_ID/$CONTAINER_ID.pid.exitcode" might be different from the actual exit code?

Reply:
That's fine, the container statemachine might be a little different, but RM won't care that.